File: src/main/java/sh/twenty/octane/commands/GmspCommand.java
package sh.twenty.octane.commands;

import org.bukkit.GameMode;
import sh.twenty.octane.Main;
import org.bukkit.entity.Player;
import sh.twenty.octane.utils.ChatUtil;


public class GmspCommand extends PlayerCommand {
    public GmspCommand(Main plugin) {super(plugin, "octane.gmsp"); }

    @Override
    public boolean execute(Player player, String[] args){
        if(args.length >= 1){
            if(!player.hasPermission("octane.gmsp.others")){
                String noPerm = plugin.getConfigManager().getMessages().getString("no-permission", "&cNo permission!");
                player.sendMessage(ChatUtil.color(noPerm));
                return false;
            }

            Player target = plugin.getServer().getPlayerExact(args[0]);

            if(target==null){
                String notFound = plugin.getConfigManager().getMessages().getString("player-not-found", "&cCould not find player: %player%");
                sendPrefixedMessage(player, notFound.replace("%player%", args[0]));
                return true;
            }

            target.setGameMode(GameMode.SPECTATOR);

            String successOther = plugin.getConfigManager().getMessages().getString("gmsp-success-other", "&7%player%'s gamemode was set to Spectator.");
            sendPrefixedMessage(player, successOther.replace("%player%", target.getName()));

            String successTarget = plugin.getConfigManager().getMessages().getString("gmsp-success-target", "&7%player%'s gamemode was set to Spectator.");
            sendPrefixedMessage(player, successTarget.replace("%player%", target.getName()));
            return true;
        }
        player.setGameMode(GameMode.SPECTATOR);

        String successSelf = plugin.getConfigManager().getMessages().getString("gmsp-success-self", "&7Your gamemode was set to Spectator");
        sendPrefixedMessage(player, successSelf);
        return true;
    }
}

File: src/main/kotlin/sh/twenty/octane/commands/GmspCommand.kt
package sh.twenty.octane.commands

import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerArgument
import dev.jorel.commandapi.kotlindsl.playerExecutor
import org.bukkit.GameMode
import org.bukkit.entity.Player
import sh.twenty.octane.OctanePlugin

class GmspCommand(private val plugin: OctanePlugin) {
    fun register() = commandAPICommand("gmsp") {
        withPermission("octane.gmsp")
        playerArgument("target", optional = true)
        playerExecutor { player, args ->
            val msgs = plugin.configManager.messages
            val target = args["target"] as? Player

            if (target != null && !player.hasPermission("octane.gmsp.others")) {
                player.sendRichMessage(msgs.prefix + msgs.noPermission)
                return@playerExecutor
            }

            val finalTarget = target ?: player
            finalTarget.gameMode = GameMode.SPECTATOR

            if (target != null) {
                player.sendRichMessage(msgs.prefix + msgs.gmspOther.replace("%player%", finalTarget.name))
                finalTarget.sendRichMessage(msgs.prefix + msgs.gmspTarget.replace("%sender%", player.name))
            } else {
                player.sendRichMessage(msgs.prefix + msgs.gmspSelf)
            }
        }
    }
}
